Add option to disable clients in netapi

barneysowood opened the pull request:
### What does this PR do?

Adds a new config option "netapi_disable_clients" that takes a list of clients to disable in the netapi.

Checks the list early in handling the the request before authentication occurs. Should be useful where certain clients (eg ssh or wheel) aren't required and should be disabled to reduce attack surface in the salt-api.

replaces https://github.com/saltstack/salt/pull/58872

### What issues does this PR fix or reference?


### New Behavior

Adds "netapi_disable_clients" list option to the config which is checked when a request is passed to NetApiClient.run() and an exception raised if the requested client is in the list.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs
- [X]  Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
No

barneysowood committed:
Add option to disable clients in netapi

Adds an option to allow you to disable clients (eg ssh, wheel) in the
netapi. Does the check before any attempts to authenticate.

barneysowood committed:
Add changelog entry

barneysowood committed:
Add documentation for netapi_disable_client option

barneysowood commented:
re-run ci/py3/macosxmojave/pytest

barneysowood commented:
re-run pr-macosxmojave-py3-pytest

sagetherage commented:
@barneysowood I updated the branch with `master` as I think we have fixed these tests that are failing, here. That reruns the full test suite so I will check back, as well. I hope that that helps!

barneysowood commented:
> @barneysowood I updated the branch with `master` as I think we have fixed these tests that are failing, here. That reruns the full test suite so I will check back, as well. I hope that that helps!

Thanks @sagetherage. Looks like the failing tests are fixed now - thanks! Unfortunately seems to have times out on ci/py3/centos7/pycryptodome/pytest - I'll try that again now.

barneysowood commented:
re-run pr-centos7-py3-pycryptodome-pytest

barneysowood commented:
OK, tests now all green.

@garethgreenaway - if you get a chance to review that would be great, thanks!

Ch3LL commented:
I approved but look like there is a merge conflict.  Would also like to get @dwoz review here as well.

barneysowood commented:
Hmm, have totally messed up the rebase trying to sort the conflict. Will fix that and re-push

barneysowood commented:
re-run pr-freebsd-122-amd64-py3-pytest

barneysowood commented:
re-run pr-ubuntu-2004-arm64-py3-pytest

barneysowood commented:
re-run pr-macosx-mojave-x86_64-py3-pytest

barneysowood commented:
@Ch3LL and @dwoz - merge conflict resolved and all tests now passing. If you could review that would be great, thanks.

barneysowood commented:
re-run pr-fedora-34-x86_64-py3-pytest

barneysowood commented:
re-run pr-fedora-34-x86_64-py3-pytest

barneysowood commented:
re-run pr-macosx-catalina-x86_64-py3-pytest

barneysowood commented:
re-run pr-macosx-mojave-x86_64-py3-pytest

dwoz reviewed (changes requested):
Honestly, due to the amount of problems we've had with the netapi I think this should be an opt in scenario rather than opt out. Meaning it'd be `netapi_enable_clients` and they are all disabled by default.

barneysowood commented:
> Honestly, due to the amount of problems we've had with the netapi I think this should be an opt in scenario rather than opt out. Meaning it'd be `netapi_enable_clients` and they are all disabled by default.

@dwoz - I would tend to agree, but I was trying to avoid a breaking change. If everyone is OK with that I'm happy rework this to be `netapi_enable_clients` and also add some documentation for that.

The other option would be to set the default (either using the `netapi_enable_clients` or `netapi_disable_clients` form) to disable the ssh client by default as that has tended to be the most problematic.

barneysowood commented:
re-run pr-amazon-2-x86_64-py3-pytest

Ch3LL commented:
@dwoz I really like the idea of the opt in as well, but that seems like a big change for users of salt-api. Shouldn't we add a deprecation notice somehow or give users notice before we make this change?

barneysowood committed:
Add pytest integration tests

Adds pytest based integration tests as the old non-pytest
netapi/test_client.py tests have been removed.

barneysowood commented:
Closing in favour of https://github.com/saltstack/salt/pull/63050
